plutus-core/plutus-core/src/PlutusCore/Builtin/KnownType.hs

```diff
@@ -196,17 +200,35 @@ typeMismatchError uniExp uniAct = fromString $ concat
 -- failure message and evaluation is about to be shut anyway.
 {-# NOINLINE typeMismatchError #-}
 
+{- Note [MakeKnownM and ReadKnownM being type synonyms]
+Normally it's a good idea for an exported abstraction not to be a type synonym, since a @newtype@
+is cheap, looks good in error messages and clearly emphasize an abstraction barrier. However we
+make 'MakeKnownM' and 'ReadKnownM' type synonyms for convenience: that way we don't need to derive
+a ton of instances (and add new ones whenever we need them), wrap and unwrap all the time
+(including in user code), which can be non-trivial for such performance-sensitive code (see e.g.
+'coerceVia' and 'coerceArg') and there is no abstraction barrier anyway.
+-}
+
+-- See Note [MakeKnownM and ReadKnownM being type synonyms].
+-- | The monad that 'makeKnown' runs in.
+type MakeKnownM cause = ExceptT (ErrorWithCause KnownTypeError cause) Emitter
```

Comment:
Is there an existing note where we explain that we only allow emitting effects at the end?

Reply:
This is described in the docs above the default sets of builtins, plus attempting to use `Emitter` in argument position gives you a nice custom type error saying you can't do that.

Technically speaking... For deferred unlifting we could as well allow `Emitter` in argument position due to unlifting happening in the same place where lifting happens anyway. I'm not sure if that could be useful and it most likely would slow things down a bit (unless! We give up on `transformers`, which is something that I'm going to try anyway).